File: 9week/남규/1. Reconstruct Itinerary.cpp
class Solution {
public:
    int ticketCount = 0;
    unordered_map<string, unordered_map<string, int>> fromTo;
    unordered_map<string, vector<string>> adj;
    
    vector<string> ans;
    
    void dfs(string from, vector<string>& path){
        if(ans.size() != 0) return;
        
        if(ticketCount == 0){
            ans = path;
            return;
        }
        
        for(int i = 0; i < adj[from].size(); i++){
            string &to = adj[from][i];
            if(fromTo[from][to] > 0){
                path.push_back(to);
                fromTo[from][to]--;
                ticketCount--;
                dfs(to, path);
                fromTo[from][to]++;
                ticketCount++;
                path.pop_back();
            }
        }
    }
    
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        for(int i = 0; i < tickets.size(); i++){
            if(tickets[i][0] == tickets[i][1]) continue;
            fromTo[tickets[i][0]][tickets[i][1]]++;
            adj[tickets[i][0]].push_back(tickets[i][1]);
            ticketCount++;
        }
        
        for(auto &v : adj){
            sort(v.second.begin(), v.second.end());
        }

        vector<string> path = {"JFK"};
        dfs("JFK", path);
        return ans;
    }
};

File: 9week/홍준/1.reconstruct-itinerary.py
## https://leetcode.com/problems/reconstruct-itinerary/

from typing import List

class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        from collections import defaultdict
        self.flightMap = defaultdict(list)

        for ticket in tickets:
            origin, dest = ticket[0], ticket[1]
            self.flightMap[origin].append(dest)

        # sort the itinerary based on the lexical order
        for origin, itinerary in self.flightMap.items():
        # Note that we could have multiple identical flights, i.e. same origin and destination.
            itinerary.sort(reverse=True)

        self.result = []
        self.DFS('JFK')

        # reconstruct the route backwards
        return self.result[::-1]

    def DFS(self, origin):
        destList = self.flightMap[origin]
        while destList:
            #while we visit the edge, we trim it off from graph.
            nextDest = destList.pop()
            print("now : " + origin + ", " + nextDest)
            self.DFS(nextDest)
        print("append!" + origin)
        self.result.append(origin)

# print(Solution().findItinerary([["MUC","LHR"],["JFK","MUC"],["SFO","SJC"],["LHR","SFO"]]))
print(Solution().findItinerary([["JFK","SFO"],["JFK","ATL"],["SFO","ATL"],["SFO","JFK"],["ATL","SFO"]]))